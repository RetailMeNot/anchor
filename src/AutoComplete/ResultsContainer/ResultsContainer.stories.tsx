// REACT
import * as React from 'react';
// STORYBOOK
import { storiesOf } from '@storybook/react';
// import { text } from '@storybook/addon-knobs';
// VENDOR
import styled from 'styled-components';
// COMPONENTS
import ResultsContainer from './ResultsContainer.component';
// README
import * as README from './README.md';
// THEME

const StyledStory = styled.div`
    padding: 2rem 5rem;
`;

storiesOf('Components/AutoComplete/ResultsContainer', module)
    .addParameters({
        readme: {
            sidebar: README,
        },
    })
    .add('Default', () => (
        <StyledStory>
            <p>ResultsContainer 1</p>
            <ResultsContainer
                emitSelectedItem={() => true}
                emitActiveIndex={() => true}
                emitActiveTerm={() => true}
                dataSource={[]}
            />
        </StyledStory>
    ));
