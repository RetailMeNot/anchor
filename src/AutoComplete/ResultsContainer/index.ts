export { ResultsContainer } from './ResultsContainer.component';
