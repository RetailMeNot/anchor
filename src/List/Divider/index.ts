export { Divider } from './Divider.component';
