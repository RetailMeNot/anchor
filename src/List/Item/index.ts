export { Item } from './Item.component';
