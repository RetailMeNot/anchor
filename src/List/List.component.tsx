// REACT
import * as React from 'react';
// VENDOR
import classNames from 'classnames';
import styled from 'styled-components';
// COMPONENTS
// THEME
import { fonts, colors, sizes } from '../theme';

interface ListProps {
  className?: string;
  children?: any;
  title?: string;
  hidden?: boolean;
}

export const StyledList = styled.div`
  .title {
    margin: 0;
    padding: ${sizes.padding.md};
    font-family: ${fonts.fontFamily};
    font-size: .75rem;
    font-weight: 600;
    color: ${colors.neutrals.ash.dark};
    line-height: .75rem;
  }
`;

const DefaultProps: ListProps = {
};

export const List: React.SFC<ListProps> = ({
  className,
  children,
  title,
...props
}: ListProps = DefaultProps): React.ReactElement<any> => (
  <StyledList className={classNames(className)} {...props}>
    {title && <p className="title">{title}</p>}
    {children}
  </StyledList>
);

List.defaultProps = DefaultProps;

export default List;
