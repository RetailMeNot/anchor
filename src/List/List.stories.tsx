// REACT
import * as React from 'react';
// STORYBOOK
import { storiesOf } from '@storybook/react';
// VENDOR
import styled from 'styled-components';
// COMPONENTS
import { List, Item } from './';
// THEME

const StyledStory = styled.div`
  padding: 2rem 5rem;
`;

storiesOf('Components/List', module)
  .add('Default', () => (
      <StyledStory>
          <p>List</p>
          <List>
            <Item>Item 1</Item>
            <Item>Item 3</Item>
            <Item>Item 4</Item>
            <Item>Item 5</Item>
            <Item>Item 6</Item>
            <Item>Item 7</Item>
          </List>
      </StyledStory>
  ));
