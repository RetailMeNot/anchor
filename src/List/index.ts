export { default as List } from './List.component';
export { default as Item } from './Item/Item.component';
