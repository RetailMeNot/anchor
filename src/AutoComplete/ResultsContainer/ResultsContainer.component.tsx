// REACT
import * as React from 'react';
// VENDOR
import classNames from 'classnames';
import styled, { css } from '@xstyled/styled-components';
import { th } from '@xstyled/system';
import { List } from '../../List';
// UTILS
const { useState, forwardRef, useImperativeHandle } = React;

type DataItem = {
    value: any;
    label: string;
    [key: string]: any;
};

type EmitSelectedItem = (newItem: DataItem) => void;
type EmitActiveTerm = (newTerm: string) => void;

interface ResultsContainerProps extends React.HTMLAttributes<HTMLDivElement> {
    initialIndex?: number;
    className?: string;
    size?: 'sm' | 'md' | 'lg';
    term: string;
    dataSource: any[];
    emitSelectedItem: EmitSelectedItem;
    emitActiveTerm: EmitActiveTerm;
    resultTemplate?: (props: ResultItemProps) => any;
}
interface StyledResultsContainerProps {
    children?: any;
    size?: 'sm' | 'md' | 'lg';
}

export interface ResultItemProps {
    term: string;
    currentIndex: number;
    index: number;
    label: string;
    value: any;
}

const ResultContainerSpaceFromAutoComplete = {
    sm: '2.6rem',
    md: '3.25rem',
    lg: '3.25rem',
};

const StyledResultsContainer = styled('div')<StyledResultsContainerProps>`
    background-color: white;
    position: absolute;
    width: inherit;
    z-index: 3;
    box-sizing: border-box;
    box-shadow: 0 0.5rem 0.75rem -0.375rem rgba(0, 0, 0, 0.2);
    border-radius: 0 0 ${th.radius('base')} ${th.radius('base')};
    padding: 1rem;
    ${({ size = 'md' }) =>
        css({ top: ResultContainerSpaceFromAutoComplete[size] })};
`;

const createResult = (label: string, value?: any): DataItem => ({
    label: label,
    value: value || label,
});

type ObjectItem = { label: string; [key: string]: any };

const generateResults = (
    results: any[],
    currentIndex: number,
    setCurrentIndex: any,
    emitSelectedItem: EmitSelectedItem,
    term: string,
    setTerm: any
): any => {
    const cleanResults: DataItem[] = [];
    if (Array.isArray(results) && results.length) {
        // Generic arrays are converted to DataItems
        if (typeof results[0] === 'string') {
            results.forEach((result: string, index: number) => {
                const itemIndex = relativeIndex(index);
                cleanResults.push(
                    createResult(result, {
                        label: result || '',
                        // Add one to index so that the input is an index
                        active: itemIndex === currentIndex,
                        key: `anchor-result-${itemIndex}`,
                        onMouseOver: () => {
                            setCurrentIndex(itemIndex);
                            setTerm(term);
                        },
                        onSelect: () =>
                            emitSelectedItem({
                                label: result,
                                value: { label: result, value: result },
                            }),
                    })
                );
            });
        } else if (typeof results[0] === 'object') {
            results.forEach(
                ({ label = '', ...props }: ObjectItem, index: number) => {
                    const itemIndex = relativeIndex(index);
                    cleanResults.push(
                        createResult(label, {
                            label,
                            ...props,
                            // Add one to index so that the input is an index
                            active: itemIndex === currentIndex,
                            key: `anchor-result-${itemIndex}`,
                            onMouseOver: () => {
                                setCurrentIndex(itemIndex);
                                setTerm(term);
                            },
                            onSelect: () =>
                                emitSelectedItem({
                                    label,
                                    value: { label, ...props },
                                }),
                        })
                    );
                }
            );
        }
    }
    return cleanResults;
};

/*
    Entire List:    [ 0, 1, 2, 3, 4 ]
    Iterable List:  [ 1, 3, 4 ]

    forward:
        - go to next iterable index
        - unless at max; go to 0 (assign initialTerm)
    backward:
        - go to previous iterable index
        - unless at 0; go to max (assign initialTerm)
    both
        - store initialTerm
        - Emit initialTerm or currentLabel
        - Update currentIndex
    ===================================================

    getNext(currentTerm: string, max = iterableList.length - 1, min = 0, currentIndex, direction): number
    emitTermAndUpdateIndex(newIndex, values): void
        • updateIndex(emitActiveIndex, setCurrentIndex): void
        • updateTerm(emitActiveIndex, setCurrentIndex): void
*/

const relativeIndex = (index: number) => index + 1;

export const ResultsContainer = forwardRef(
    (
        {
            className,
            dataSource = [],
            emitSelectedItem,
            emitActiveTerm,
            term,
            resultTemplate,
            ...props
        }: ResultsContainerProps,
        resultsContainerRef: React.Ref<any>
    ): JSX.Element => {
        const [currentIndex, setCurrentIndex] = useState<number>(0);
        const [initialTerm, setInitialTerm] = useState<string>('');

        const results = generateResults(
            dataSource,
            currentIndex,
            setCurrentIndex,
            emitSelectedItem,
            term,
            setInitialTerm
        );

        const iterativeIndexes: number[] = [0];

        results.forEach(({ value: { listItemType } }: any, index: number) => {
            if (!listItemType || listItemType === 'item') {
                // Add one to index to allow input to be item 0
                iterativeIndexes.push(relativeIndex(index));
            }
        });

        const traverse = (
            range: any[],
            index: number,
            forward: boolean
        ): number => {
            // Get the max range; the minimum index of any array is 0
            const max = range.length - 1;
            // Determine the actual index relative to all iterable values
            const indexRelative: number = range.indexOf(index);
            // Increment/decrement accordingly
            let nextIndex = forward ? indexRelative + 1 : indexRelative - 1;
            // For 'next' iteration, reset to 0 if at max
            if (forward && nextIndex > max) {
                nextIndex = 0;
            }
            // For 'prev' iteration, reset to max if at 0
            if (!forward && nextIndex < 0) {
                nextIndex = max;
            }
            return nextIndex;
        };

        const handleTraversal = (
            currentTerm: string = '',
            forward: boolean
        ) => {
            // Check if there's a initialTerm
            if (!initialTerm || initialTerm === '') {
                // If there's no initialTerm, assign the current input value
                setInitialTerm(currentTerm);
            }
            // Determine the next index
            const nextIndex = traverse(iterativeIndexes, currentIndex, forward);
            // Assign that index locally
            setCurrentIndex(iterativeIndexes[nextIndex]);
            if (nextIndex === 0) {
                // Emit the user's initial term
                emitActiveTerm(initialTerm);
                setInitialTerm('');
            } else {
                // Emit the active term, but subtract one bc the input is not part of the original data set
                emitActiveTerm(results[iterativeIndexes[nextIndex] - 1].label);
            }
        };

        useImperativeHandle(resultsContainerRef, () => ({
            setActiveIndex: (itemIndex: number) => {
                setCurrentIndex(itemIndex);
            },
            clearInitialTerm: () => {
                setInitialTerm('');
            },
            handleNext: (currentTerm: string) => {
                handleTraversal(currentTerm, true);
            },
            handlePrevious: (currentTerm: string) => {
                handleTraversal(currentTerm, false);
            },
            selectActive: () => {
                if (currentIndex >= 0 && currentIndex <= results.length) {
                    emitSelectedItem(results[currentIndex - 1]);
                }
            },
        }));
        return (
            <StyledResultsContainer
                className={classNames(
                    'anchor-auto-complete-results-container',
                    className
                )}
                {...props}
            >
                {resultTemplate ? (
                    <div className="auto-complete-results">
                        {results.map(
                            (
                                { label, value }: ResultItemProps,
                                index: number
                            ) =>
                                React.createElement(resultTemplate, {
                                    label,
                                    value,
                                    term,
                                    currentIndex,
                                    index: relativeIndex(index),
                                    key: relativeIndex(index),
                                })
                        )}
                    </div>
                ) : (
                    <List items={results} className="auto-complete-results" />
                )}
            </StyledResultsContainer>
        );
    }
);
