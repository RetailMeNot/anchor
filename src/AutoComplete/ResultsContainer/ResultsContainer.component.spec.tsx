// REACT
import * as React from 'react';
import * as renderer from 'react-test-renderer';
// COMPONENT
import { ResultsContainer } from './ResultsContainer.component';
import { Item } from '../../List';
// ENZYME
import { shallow, mount } from 'enzyme';

describe('Component: ResultsContainer', () => {
    it('should be defined', () => {
        const subject = (
            <ResultsContainer
                term="a"
                emitSelectedItem={() => true}
                emitActiveTerm={() => true}
                dataSource={[
                    { value: 1, label: 'Result Item 1' },
                    { value: 2, label: 'Result Item 2' },
                ]}
            />
        );
        const wrapper = mount(subject);
        const component = shallow(subject);

        expect(subject).toBeDefined();
        expect(wrapper).toBeDefined();
        expect(component).toBeDefined();

        const tree = renderer.create(subject).toJSON();
        expect(tree).toMatchSnapshot();
    });

    it('should accept a resultTemplate', () => {
        const CustomResult = ({ index, currentIndex, label }: any) => (
            <Item key={index} active={index === currentIndex}>
                {label}
            </Item>
        );

        const subject = (
            <ResultsContainer
                term="a"
                emitSelectedItem={() => true}
                emitActiveTerm={() => true}
                dataSource={[
                    { value: 1, label: 'Result Item 1' },
                    { value: 2, label: 'Result Item 2' },
                ]}
                resultTemplate={CustomResult}
            />
        );

        const tree = renderer.create(subject).toJSON();
        expect(tree).toMatchSnapshot();
    });
});
