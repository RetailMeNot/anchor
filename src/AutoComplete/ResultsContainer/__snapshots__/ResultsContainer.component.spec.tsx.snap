// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: ResultsContainer should accept a resultTemplate 1`] = `
.c1 {
  box-sizing: border-box;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  border-radius: 4px;
  padding: 0.5rem 1rem;
  cursor: pointer;
  -webkit-transition: background 500ms;
  transition: background 500ms;
  background: white;
  -webkit-text-decoration: none;
  text-decoration: none;
  height: 2.75rem;
}

.c1:hover {
  background: #FAFAFA;
}

.c1.active {
  background: #F1F1F1;
}

.c3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-content: space-between;
  -ms-flex-line-pack: space-between;
  align-content: space-between;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  width: 100%;
  height: 100%;
}

.c3 .item-prefix {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-flex: 0 0 0;
  -ms-flex: 0 0 0;
  flex: 0 0 0;
}

.c3 .item-main {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-flex: 4 1 auto;
  -ms-flex: 4 1 auto;
  flex: 4 1 auto;
}

.c3 .item-suffix {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-flex: 0 0 0;
  -ms-flex: 0 0 0;
  flex: 0 0 0;
  text-align: right;
  -webkit-box-pack: end;
  -webkit-justify-content: flex-end;
  -ms-flex-pack: end;
  justify-content: flex-end;
}

.c3.lg {
  font-size: 1rem;
  font-weight: 400;
}

.c3.sm {
  font-size: 0.875rem;
  font-weight: 500;
}

.c0 {
  background-color: white;
  position: absolute;
  width: inherit;
  z-index: 3;
  box-sizing: border-box;
  box-shadow: 0 0.5rem 0.75rem -0.375rem rgba(0,0,0,0.2);
  border-radius: 0 0 4px 4px;
  padding: 1rem;
  top: 3.25rem;
}

.c2 {
  box-sizing: border-box;
  font-family: Avenir Next,Segoe UI,Helvetica Neue,Helvetica,Roboto,sans-serif;
  margin: 0;
  font-size: 16px;
  line-height: 1.5rem;
  font-weight: normal;
  text-align: inherit;
  color: inherit;
  color: #595959;
}

.c2 small {
  font-size: 87.5%;
}

.c4 {
  box-sizing: border-box;
  font-family: Avenir Next,Segoe UI,Helvetica Neue,Helvetica,Roboto,sans-serif;
  margin: 0;
  font-size: 16px;
  line-height: 1.5rem;
  font-weight: normal;
  text-align: inherit;
  color: inherit;
  color: #595959;
}

.c4 small {
  font-size: 87.5%;
}

<div
  className="c0 anchor-auto-complete-results-container"
>
  <div
    className="auto-complete-results"
  >
    <a
      className="c1 anchor-list-item"
      onMouseDown={[Function]}
      onMouseOut={[Function]}
      onMouseOver={[Function]}
    >
      <span
        className="c2 anchor-typography c3 lg"
      >
        Result Item 1
      </span>
    </a>
    <a
      className="c1 anchor-list-item"
      onMouseDown={[Function]}
      onMouseOut={[Function]}
      onMouseOver={[Function]}
    >
      <span
        className="c4 anchor-typography c3 lg"
      >
        Result Item 2
      </span>
    </a>
  </div>
</div>
`;

exports[`Component: ResultsContainer should be defined 1`] = `
.c2 {
  box-sizing: border-box;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  border-radius: 4px;
  padding: 0.5rem 1rem;
  cursor: pointer;
  -webkit-transition: background 500ms;
  transition: background 500ms;
  background: white;
  -webkit-text-decoration: none;
  text-decoration: none;
  height: 2.75rem;
}

.c2:hover {
  background: #FAFAFA;
}

.c2.active {
  background: #F1F1F1;
}

.c4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-content: space-between;
  -ms-flex-line-pack: space-between;
  align-content: space-between;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  width: 100%;
  height: 100%;
}

.c4 .item-prefix {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-flex: 0 0 0;
  -ms-flex: 0 0 0;
  flex: 0 0 0;
}

.c4 .item-main {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-flex: 4 1 auto;
  -ms-flex: 4 1 auto;
  flex: 4 1 auto;
}

.c4 .item-suffix {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-flex: 0 0 0;
  -ms-flex: 0 0 0;
  flex: 0 0 0;
  text-align: right;
  -webkit-box-pack: end;
  -webkit-justify-content: flex-end;
  -ms-flex-pack: end;
  justify-content: flex-end;
}

.c4.lg {
  font-size: 1rem;
  font-weight: 400;
}

.c4.sm {
  font-size: 0.875rem;
  font-weight: 500;
}

.c1 {
  box-sizing: border-box;
}

.c1 .anchor-list-title {
  margin-top: 0.75rem;
}

.c1 .anchor-list-item {
  margin-bottom: 0.25rem;
}

.c1 .anchor-list-item + .anchor-list-title {
  margin-top: 0.5rem;
}

.c0 {
  background-color: white;
  position: absolute;
  width: inherit;
  z-index: 3;
  box-sizing: border-box;
  box-shadow: 0 0.5rem 0.75rem -0.375rem rgba(0,0,0,0.2);
  border-radius: 0 0 4px 4px;
  padding: 1rem;
  top: 3.25rem;
}

.c3 {
  box-sizing: border-box;
  font-family: Avenir Next,Segoe UI,Helvetica Neue,Helvetica,Roboto,sans-serif;
  margin: 0;
  font-size: 16px;
  line-height: 1.5rem;
  font-weight: normal;
  text-align: inherit;
  color: inherit;
  color: #595959;
}

.c3 small {
  font-size: 87.5%;
}

.c5 {
  box-sizing: border-box;
  font-family: Avenir Next,Segoe UI,Helvetica Neue,Helvetica,Roboto,sans-serif;
  margin: 0;
  font-size: 16px;
  line-height: 1.5rem;
  font-weight: normal;
  text-align: inherit;
  color: inherit;
  color: #595959;
}

.c5 small {
  font-size: 87.5%;
}

<div
  className="c0 anchor-auto-complete-results-container"
>
  <div
    className="c1 anchor-list auto-complete-results"
  >
    <a
      className="c2 anchor-list-item"
      label="Result Item 1"
      onMouseDown={[Function]}
      onMouseOut={[Function]}
      onMouseOver={[Function]}
    >
      <span
        className="c3 anchor-typography c4 lg"
      >
        Result Item 1
      </span>
    </a>
    <a
      className="c2 anchor-list-item"
      label="Result Item 2"
      onMouseDown={[Function]}
      onMouseOut={[Function]}
      onMouseOver={[Function]}
    >
      <span
        className="c5 anchor-typography c4 lg"
      >
        Result Item 2
      </span>
    </a>
  </div>
</div>
`;
